p2p/src/sync/mod.rs

```diff
@@ -488,46 +173,110 @@ where
             .await
             .map_err(|_| P2pError::SubsystemFailure)?;
 
-        Ok(rx)
+        Ok(receiver)
     }
 
-    async fn process_block_announcement(
-        &mut self,
-        peer_id: T::PeerId,
-        block: Block,
-    ) -> crate::Result<()> {
-        let result = match self
+    // TODO: This shouldn't be public.
+    /// Registers the connected peer by creating a context for it.
+    ///
+    /// The `HeaderListRequest` message is sent to newly connected peers.
+    pub async fn register_peer(&mut self, peer: T::PeerId) -> Result<()> {
+        log::debug!("Register peer {peer} to sync manager");
+
+        self.request_headers(peer).await?;
+        match self.peers.insert(peer, PeerContext::new()) {
+            // This should never happen because a peer can only connect once.
+            Some(_) => Err(P2pError::PeerError(PeerError::PeerAlreadyExists)),
+            None => Ok(()),
+        }
+    }
+
+    /// Removes the state (`PeerContext`) of the given peer.
+    fn unregister_peer(&mut self, peer: T::PeerId) {
+        log::debug!("Unregister peer {peer} from sync manager");
+
+        // Remove the queued block responses associated with the disconnected peer.
+        self.blocks_queue.retain(|(p, _)| p != &peer);
+
+        if self.peers.remove(&peer).is_some() {
+            log::warn!("Unregistering unknown peer: {peer}");
+        }
+    }
+
+    /// Announces the header of a new block to peers.
+    async fn handle_new_tip(&mut self, block_id: Id<Block>) -> Result<()> {
+        if self.is_initial_block_download {
+            self.is_initial_block_download =
+                self.chainstate_handle.call(|c| c.is_initial_block_download()).await??;
```

Comment:
Why call twice, once for is_initial_block_download() and once for get_block()? Why not both in one call?